libr/util/str.c

```diff
@@ -10,6 +10,78 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+R_API RCharset *r_charset_new(void) {
+	RCharset* rcs = R_NEW0 (RCharset);
+	if (!rcs) {
+		return NULL;
+	}
+	rcs->custom_charset = NULL;
+	rcs->remaining = 0;
+	return rcs;
+}
+
+R_API RCharsetRune *r_charset_rune_new(const ut8 *ch, const ut8 *hx) {
+	RCharsetRune* rcr = R_NEW0 (RCharsetRune);
+	if (!rcr) {
+		return NULL;
+	}
+	rcr->ch = (ut8 *) strdup ((char *) ch);
+	rcr->hx = (ut8 *) strdup ((char *) hx);
+	rcr->left = NULL;
+	rcr->right = NULL;
+	return rcr;
+}
+
+R_API void r_charset_rune_free(RCharsetRune *rcr) {
+	free (rcr->ch);
+	free (rcr->hx);
+	free (rcr);
+}
+
+R_API RCharsetRune *add_rune(RCharsetRune *rcsr, const ut8 *ch, const ut8 *hx) {
+	if (rcsr == NULL) {
+		rcsr = r_charset_rune_new (ch, hx);
+	}
+
+	if (strcmp ((char *)hx, (char *)rcsr->hx) < 0) {
+		rcsr->left = add_rune (rcsr->left, ch, hx);
+	} else if (strcmp ((char *)hx, (char *)rcsr->hx) > 0) {
+		rcsr->right = add_rune (rcsr->right, ch, hx);
+	} else {
+		if(strcmp ((char *)ch, (char *)rcsr->ch) > 0) {
+			rcsr->left = add_rune (rcsr->left, ch, hx);
+		} else if(strcmp ((char *)ch, (char *)rcsr->ch) < 0) {
+			rcsr->right = add_rune (rcsr->right, ch, hx);
+		}
+	}
+
+	return rcsr;
+}
+
+R_API RCharsetRune *search_from_hex(RCharsetRune *rcsr, const ut8 *hx) {
+	if (!rcsr) {
+		return NULL;
+	}
+	else if (strcmp ((char *)rcsr->hx, (char *)hx) == 0) {
+		return rcsr;
+	}
+	else {
+		RCharsetRune *left = search_from_hex(rcsr->left, hx);
+		return left? left:  search_from_hex(rcsr->right, hx);
+	}
+}
+
+R_API RCharsetRune *search_from_char(RCharsetRune *rcsr, const ut8 *ch) {
+	if (rcsr == NULL)
+	    return NULL;
+	else if ( strcmp ((char *)rcsr->ch, (char *)ch) == 0)
```

Comment:
dont else after return

Reply:
use ! instead of == 0'

Reply:
ok for the `!`.  I do not understand for the else. The else is to search to left or right in the search tree.

Reply:
If you return the else statement is not necessary because nothing else can follow

Reply:
yes but if the condition is not filled? it is an else if. not an else.